### PR TITLE
feat: PyTorch Trainer API

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

currently the training loop + dtrain implementation. features like profiling, AMP, checkpointing, etc still in progress.

## Checklist
- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: harness/determined/pytorch/trainer.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+            dist.init_process_group(backend="gloo")  # type: ignore
+        return core.DistributedContext.from_torch_distributed()
+    else:
+        print(f"Backend {distributed_backend} not supported")
+
+
+@contextlib.contextmanager
+def init():
+    distributed_context = initialize_distributed_backend()
+    core_context = core.init(distributed=distributed_context)
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/trainer.py:282]: this needs to be `with core.init(...) as core_context`.

The way it's written right now, `core_context.close()` doesn't get ever called.

Comment (rb-determined-ai) [reply harness/determined/pytorch/trainer.py:282]: I'd say at least 90% of the time, a contextmanager that doesn't have a `try: yield; finally: ...` or a `with X: yield` is not doing what you want, because if you are using a contextmanager it is so that you can do some cleanup step after execution continues from `yield`.

There's like 10% of the time where the steps you would complete afterwards should only be completed in success cases, and that's the only time you have neither a `try` nor a `with`.

File: harness/determined/__init__.py [hunk 2]
```diff
@@ -17,6 +17,7 @@
     InvalidHP,
 )
 from determined import errors
+from determined import pytorch
```
Comment (rb-determined-ai) [inline harness/determined/__init__.py:20]: no adding mandatory frameworks to determined.  This would be very annoying to pure tensorflow users, or pure core api users who might be using something like scikit learn and neither pytorch nor tensorflow.

File: harness/determined/_trial_context.py [hunk 3]
```diff
@@ -14,10 +14,12 @@ class TrialContext:
     def __init__(
         self,
         core_context: core.Context,
-        env: det.EnvContext,
+        env: Optional[det.EnvContext] = None,
```
Comment (rb-determined-ai) [inline harness/determined/_trial_context.py:17]: how do you feel about removing the `env` object entirely?

It's basically a pure tech debt item.  In local training mode, it's full of phony values that we make up to do local training.  I would like to see the `env` object not be found anywhere in our pytorch code when the pytorch trainer API lands.

I am not saying you need to do it now before we've finish iterating on the API, of course.  But I think it's important to do in the "implementation and automated testing" milestone, because I suspect there will be things in env that end up needing to be passed in by the user in `pytorch.init()` or something... which is important we don't release an API and figure that out later.

Comment (rb-determined-ai) [reply harness/determined/_trial_context.py:17]: oh, and definitely I'm not proposing any modifications to EstimatorTrialContext or TFKerasTrialContext.  IMO those can continue taking an `env` as part of their subclass definition.

File: harness/determined/pytorch/_pytorch_context.py [hunk 4]
```diff
@@ -51,9 +51,18 @@ class PyTorchTrialContext(det.TrialContext, pytorch._PyTorchReducerContext):
     def __init__(self, *args: Any, **kwargs: Any) -> None:
         det.TrialContext.__init__(self, *args, **kwargs)
         pytorch._PyTorchReducerContext.__init__(self, self.distributed.allgather)
+
+        self.cluster_info = det.get_cluster_info()
+        if self.cluster_info:
+            self.managed_training = True
+        else:
+            self.managed_training = False
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_context.py:59]: Is this accurate?  What if I want to do local training, but I happen to be in a determined notebook or shell?

File: harness/determined/pytorch/_pytorch_context.py [hunk 5]
```diff
@@ -51,9 +51,18 @@ class PyTorchTrialContext(det.TrialContext, pytorch._PyTorchReducerContext):
...
+
+        self.experiment_config = self.cluster_info.trial._config
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_context.py:61]: Having a `self.experiment_config` on the PytorchTrialContext is sort of antithetical to the goal of supporting local training mode.

How hard/impossible is it to remove it entirely?  What would it take to get there?

File: harness/determined/_trial_context.py (hunk 3 above)
Comment (rb-determined-ai) [reply harness/determined/_trial_context.py:17]: or... it looks like you already removed all env references from pytorchtrial and pytorchcontext?  If that's the case, then you're like 90% of the way there.

Last step is make sure there's not even a `self.env` that other developers might be tempted to use, ha

File: harness/determined/pytorch/trainer.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+        val_loader = val_data.get_data_loader(
+            repeat=False, num_replicas=num_replicas, rank=rank
+        )
+        return val_loader
+
+    def train(
+        self,
+        max_epochs: Optional[int] = None,
+        # OR
+        max_batches: Optional[int] = None,
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/trainer.py:70]: why not something like:
`max_length=Epochs(10)` or `max_length=Batches(10000)`?

File: harness/determined/pytorch/trainer.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        average_aggregated_gradients: bool = True,
+        aggregation_frequency: int = 2,
+        searcher_metric="validation_loss",
+        profiling=True,
+        profiling_start=0,
+        profiling_end=10,
+        sync_timings=None,
+        checkpoint_policy="best|all|none"
+    ):
+        assert (max_epochs is None) ^ (max_batches is None), "Either max_batches or max_epochs must be defined"
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/trainer.py:83]: neat use of xor, I approve

File: harness/determined/pytorch/trainer.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+            max_batches,
+            min_validation_period,
+            min_checkpoint_period,
+        )
+
+        train_loop.run()
+
+        return
+
+    def on_training_epoch_start(self, epoch_idx: int):
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/trainer.py:118]: no private helper functions with public-looking names, please use an underscore prefix.
